### PR TITLE
Add support for logaddexp(float16) in CUDA and implement its reference

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91869

The reference is implemented so that it generates efficient and
numerically stable triton code.

Fixes https://github.com/pytorch/pytorch/issues/91683

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by lezcano** [comment]
The implementation is slightly better than that we have in CUDA in master I think.

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -1432,6 +1432,20 @@ def le(a: TensorLikeType, b: TensorLikeType) -> TensorLikeType:
...
 
+@_make_elementwise_binary_reference(
+    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+    supports_lhs_python_scalar=False,
+    supports_rhs_python_scalar=False,
+)
+def logaddexp(a: TensorLikeType, b: TensorLikeType) -> TensorLikeType:
+    # Nb. this implementation does nto distribute the gradients evenly when a == b
+    mask = a >= b
+    max_ = torch.where(mask, a, b)
```
Comment (ngimel) [inline torch/_refs/__init__.py:1443]: I'm not sure 2 masking ops are better than one `abs`, but I don't think it matters much.

**Review by ngimel — APPROVED** [review]

File: torch/_refs/__init__.py (hunk 1 above)
Comment (lezcano) [reply torch/_refs/__init__.py:1443]: it's 1 masking op vs 1 abs and a neg. It should be marginally better, but yeah

Comment (lezcano) [reply torch/_refs/__init__.py:1445]: Also, note that nvcc and may be able to fuse the two wheres into just one flow control op.
